Add ucrt64 mingw toolchain.

angerman opened the pull request:
this is mingw/gcc/ucrt the llvm based one is called clang64 as per https://www.msys2.org/docs/environments/.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

angerman commented:
```
$ strings $(nix build -f . pkgs.pkgsCross.ucrt64.hello --no-
link --print-out-paths)/bin/hello.exe | grep '\.dll'
Winsock.dll version out of range
KERNEL32.dll
api-ms-win-crt-convert-l1-1-0.dll
api-ms-win-crt-environment-l1-1-0.dll
api-ms-win-crt-filesystem-l1-1-0.dll
api-ms-win-crt-heap-l1-1-0.dll
api-ms-win-crt-locale-l1-1-0.dll
api-ms-win-crt-math-l1-1-0.dll
api-ms-win-crt-private-l1-1-0.dll
api-ms-win-crt-runtime-l1-1-0.dll
api-ms-win-crt-stdio-l1-1-0.dll
api-ms-win-crt-string-l1-1-0.dll
api-ms-win-crt-time-l1-1-0.dll
```
vs
```
$ strings $(nix build -f . pkgs.pkgsCross.mingwW64.hello --n
o-link --print-out-paths)/bin/hello.exe | grep '\.dll'
Winsock.dll version out of range
KERNEL32.dll
msvcrt.dll
```

angerman commented:
@shlevy did open #171418 a while ago, but eventually got reverted in #173487; This PR explicitly tries to _only_ do the UCRT part with GCC for now.

Ericson2314 commented:
Can you make the first commit switch all those conditionals to `isMinGW` so we don't need to enumerate both libcs, and then the second commit actually add ucrt?

angerman committed:
treewide: replace libc == "msvcrt" with isMinGW

msvcrt is only one of the libcs in MinGW. We therefore
replace explictly testing for msvcrt with the isMinGW
predicate. This lays the foundation for ucrt64 support.

angerman committed:
nixpkgs/systems: Add ucrt64 as MinGW libc

The Minimalist Gnu for Windows distribution comes with support for
the traditional msvcrt libc, as well as ucrt64 libc. The latter
being the newer universal compiler runtime. We follow the msys2
environment naming convention[1]:

| name       | toolchain | arch    | libc   | libc++    |
|------------|-----------|---------|--------|-----------|
| mingw32    | gcc       | i686    | msvcrt | libstdc++ |
| mingw64    | gcc       | x86_64  | msvcrt | libstdc++ |
| ucrt64     | gcc       | x86_64  | ucrt   | libstdc++ |
| clang32    | llvm      | i686    | ucrt   | libc++    |
| clang64    | llvm      | x86_64  | ucrt   | libc++    |
| clangarm64 | llvm      | aarch64 | ucrt   | libc++    |

For now nixpkgs only supports the first three with this commit.

--
[1]: https://www.msys2.org/docs/environments/

angerman commented:
> Can you make the first commit switch all those conditionals to `isMinGW` so we don't need to enumerate both libcs, and then the second commit actually add ucrt?

Done.

angerman commented:
Why can't I assign this to myself?

nagy commented:
Can you also please follow the commit message conventions at https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions?

angerman commented:
> Can you also please follow the commit message conventions at https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions?

@nagy Let me know if this is better. Feel free to suggest even better ones!

nagy commented:
those look ok

OPNA2608 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20687,7 +20687,7 @@ with pkgs;
   preLibcCrossHeaders = let
     inherit (stdenv.targetPlatform) libc;
-  in     if libc == "msvcrt" then targetPackages.windows.mingw_w64_headers or windows.mingw_w64_headers
+  in     if isMinGW then targetPackages.windows.mingw_w64_headers or windows.mingw_w64_headers
```
```
error: undefined variable 'isMinGW'
```

```suggestion
  in     if stdenv.targetPlatform.isMinGW then targetPackages.windows.mingw_w64_headers or windows.mingw_w64_headers
```

angerman replied:
@OPNA2608 thanks, missed that in the rewrite.

Artturin commented:
Should target master instead of a release branch.

Artturin commented on pkgs/development/compilers/gcc/10/default.nix:
The `gcc/*/default.nix` files don't exist anymore, `crossMingw` variable is now in `pkgs/development/compilers/gcc/common/configure-flags.nix`

angerman replied:
Fun, this is going to be fun back porting then.

angerman replied:
@Artturin rebased against master now.
